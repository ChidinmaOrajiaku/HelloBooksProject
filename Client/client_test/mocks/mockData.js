import { updatePassword } from "../../actions/updatePassword";

const mockData = {
  userDetails: {
    message: 'Welcome admin',
    username: 'admin',
    token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImRhZGR5Iiwicm9sZSI6InVzZXIiLCJpZCI6MiwiaWF0IjoxNTEzODc1ODE3LCJleHAiOjE1MTM5NjIyMTd9.oexUFhi08vhMU88LmZJvRDy9JFdThMpfb3Qz9DWWzgk'
  },
  userDetailsResponse: {
    username: 'daddy',
    role: 'user',
    id: 2,
    iat: 1513875817,
    exp: 1513962217
  },
  userCountResponse: {
    count: 7
  },
  borrowBookData: {
    bookId: 1,
    userId: 1,
  },
  borrowBookDataResponse: {
    returned: false,
    id: 24,
    userId: 1,
    bookId: 7,
    toReturnDate: '2018-01-14',
    updatedAt: '2017-12-21T21:16:08.442Z',
    createdAt: '2017-12-21T21:16:08.442Z',
    returnDate: null,
    categoryId: null
  },
  createBookData: {
    title: 'a',
    author: 'a',
    category: 'a',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1511133007/g2exfx6sgbymszspybts.jpg',
    review: 'Nice'
  },
  createBookErrorData: {
    author: 'a',
    category: 'a',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1511133007/g2exfx6sgbymszspybts.jpg',
    review: 'Nice'
  },
  createBookErrorResponse: 'An error occured',
  createBookResponse: {
    message: 'Succesfully added'
  },
  getABookData: {
    id: 7,
    title: 'Rage of Angels',
    author: 'Sidney Sheldon',
    category: 'Fiction',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1513096198/prbdr6eghnnrcewwpodf.jpg',
    review: 'Nice Book',
    createdAt: '2017-12-12T11:11:10.823Z',
    updatedAt: '2017-12-12T16:30:00.352Z',
    userId: null,
    categoryId: null
  },
  getAllBooksData: {
    id: 7,
    title: 'Rage of Angels',
    author: 'Sidney Sheldon',
    category: 'Fiction',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1513096198/prbdr6eghnnrcewwpodf.jpg',
    review: 'Nice Book',
    createdAt: '2017-12-12T11:11:10.823Z',
    updatedAt: '2017-12-12T16:30:00.352Z',
    userId: null,
    categoryId: null
  },
  getAllBorrowedBooksData: {
    id: 7,
    title: 'Rage of Angels',
    author: 'Sidney Sheldon',
    category: 'Fiction',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1513096198/prbdr6eghnnrcewwpodf.jpg',
    review: 'Nice Book',
    createdAt: '2017-12-12T11:11:10.823Z',
    updatedAt: '2017-12-12T16:30:00.352Z',
    userId: null,
    categoryId: null
  },
  getCategoryData: {
    category: 'Fiction'
  },
  getUserData: {
    firstname: 'daddy',
    lastname: 'daddy',
    username: 'daddy',
    password: 'daddy',
    email: 'daddy@gmail.com'
  },
  bookData: {
    title: 'a',
    author: 'b',
    category: 'a',
    image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1511133007/g2exfx6sgbymszspybts.jpg',
    review: 'Nice'
  },
  returnBookData: {
    message: 'Successfully Returned'
  },
  userData: {
    verifyPassword: '1234',
    password: '4321'
  },
  errorMessage: 'An error occurred',
  image: 'https://res.cloudinary.com/andela-chidinma/image/upload/v1515403427/ena7dhy6xriztom6vuvo.jpg'
};

export default mockData;
