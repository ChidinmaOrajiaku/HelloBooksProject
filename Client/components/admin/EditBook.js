import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import Footer from '../Footer';
import NavigationBar from '../NavigationBar';
import { saveImageCloudinary } from '../../actions/booksAction';
import { adminModifyRequest } from '../../actions/createBooks';

/**
 * @class AddBooks
 */
class AddBooks extends React.Component {
  /**
     * @constructor
     * @param {object} props 
     */
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      author: '',
      category: '',
      tempImage: '',
      image: '',
      review: '',
      pointer: false,
    };
  }

  /**
 * 
 * @returns {event} event
 * @param {any} event 
 * @memberof AddBooks
 */
  handleChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }
  /**
     * 
     * 
     * @returns {ReactElement} Markup 
     * @memberof AddBooks
     */
  render() {
    return (
      <div className="adminAddBooks row ontainer">
        <div className=""> <NavigationBar /> </div>
        <div className="col m4 offset-m5">
          <div className="card">
            <form onSubmit="" id="form">
              <div className="row">
                <div className="input-field col s12">
                  <input value={this.state.title} onChange={this.handleChange} id="title" type="text" className="validate"/>
                  <label htmlFor="title">Title</label>
                </div>
              </div>
              <div className="row">
                <div className="input-field col s12">
                  <input value={this.state.author} onChange={this.handleChange} id="author" type="text" className="validate"/>
                  <label htmlFor="author">Author</label>
                </div>
              </div>
              <div className="row">
                <div className="input-field col s12">
                  <input value={this.state.category} onChange={this.handleChange} id="category" type="text" className="validate"/>
                  <label htmlFor="category">Category</label>
                </div>
              </div>
              <div className="row">
                <div className="file-field input-field col s12">
                  <div className="btn file">
                    <span><i className="material-icons">file_upload</i></span>
                    <input type="file" onChange={this.handleImageChange} id="image"/>
                  </div>
                  <div className="file-path-wrapper">
                    <input className="file-path validate" type="text" placeholder="Upload Image"/>
                  </div>
                </div>
              </div>
              <div className="row">
                <div className="input-field col s12">
                  <textarea value={this.state.review} onChange={this.handleChange} id="review" className="materialize-textarea"></textarea>
                  <label htmlFor="review">Review</label>
                </div>
              </div>
              <button className="btn waves-effect waves-light" type="submit" name="action">Add<i className="material-icons right">send</i></button>
            </form>
          </div>
        </div>
        <div> <Footer/></div>
      </div>
    );
  }
}

const mapStateToProps = state => (
  {
    modifyBookData: state.modifyBooks[0].response,
    imageInputUrl: state.uploadImage[0].response
  }
);

export default connect(mapStateToProps, { adminModifyRequest, saveImageCloudinary })(AddBooks);
