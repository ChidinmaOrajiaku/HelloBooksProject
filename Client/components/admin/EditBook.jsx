import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Footer from '../Footer';
import NavigationBar from '../NavigationBar';
import { saveImageCloudinary } from '../../actions/booksAction';
import { adminModifyRequest } from '../../actions/modifyBooks';
import { getBookRequest } from '../../actions/getABook';
import { getAllCategoryRequest } from '../../actions/getCategory';

/**
 * @class EditBook
 */
export class EditBook extends React.Component {
  /**
     * @constructor
     * @param {object} props
     */
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      author: '',
      category: '',
      tempImage: '',
      image: '',
      imageView: '',
      imagePreview: '',
      preview: '',
      review: '',
      modified: '',
      bookData: '',
      categoryData: [],
      currentBookId: localStorage.getItem('currentBookId'),
      pointer: false,
      loading: true,
      categoryLoader: true,
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleImageChange = this.handleImageChange.bind(this);
    this.onEditCloudinaryRequest = this.onEditCloudinaryRequest.bind(this);
    this.onEditRequest = this.onEditRequest.bind(this);
  }

  /**
 *
 * @returns {object} mounted components
 * @memberof EditBook
 */
  componentDidMount() {
    this.props.getBookRequest(this.state.currentBookId);
    this.props.getAllCategoryRequest().then(() => {
      $('select').material_select();
      $('select').change(event => this.handleChange(event));
    });
    $('.modal').modal({
      dismissible: true,
      opacity: 0.5,
      inDuration: 300,
      outDuration: 200,
      startingTop: '2%',
      endingTop: '20%',
    });
  }

  /**
 *
 * @returns {nextProps} nextProps
 * @param {any} nextProps
 * @memberof EditBook
 */
  componentWillReceiveProps(nextProps) {
    if (nextProps.getABookData && this.state.loading) {
      this.setState({
        loading: false,
        bookData: nextProps.getABookData,
        title: nextProps.getABookData.title,
        author: nextProps.getABookData.author,
        category: nextProps.getABookData.category,
        imageView: nextProps.getABookData.image,
        image: nextProps.getABookData.image,
        review: nextProps.getABookData.review
      });
    }
    if (nextProps.imageInputUrl && this.state.pointer) {
      this.setState({
        image: nextProps.imageInputUrl,
        pointer: false
      }),
      setTimeout(() => {
        this.props.adminModifyRequest(this.state.currentBookId, this.state);
      }, 1000);
      setTimeout(() => {
        if (this.props.modifyBookData.isModified === true) {
          this.props.history.push('/books');
          Materialize.toast('Successfully Updated', 2000, 'teal rounded');
        } else {
          Materialize.toast('Not Updated', 2000, 'red rounded');
        }
      }, 2000);
    }

    if (nextProps.getCategoryData) {
      this.setState({
        categoryLoader: false,
        spliceCategoryData: nextProps.getCategoryData.splice(this.state.category, 1),
        categoryData: nextProps.getCategoryData
      });
    }
  }

  /**
 *
 * @returns {event} event
 * @param {any} event
 * @memberof EditBook
 */
  handleChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }

  /**
 *
 * @returns {event} handle Image change
 * @param {any} event
 * @memberof EditBook
 */
  handleImageChange(event) {
    event.preventDefault();
    const imageInput = event.target.files[0];
    const imageInputReader = new FileReader();
    if (imageInput) {
      imageInputReader.onload = () => {
        const newUpload = new Image();
        newUpload.src = imageInputReader.result;
        newUpload.onload = () => {
          this.setState({
            tempImage: imageInput,
            imageView: imageInput.name,
            imagePreview: newUpload.src
          });
        };
      };
    }
    imageInputReader.readAsDataURL(imageInput);
  }

  /**
 *
 * @returns {SyntheticEvent} event
 * @param {any} event
 * @memberof EditBook
 */
  onEditCloudinaryRequest(event) {
    event.preventDefault();
    this.props.saveImageCloudinary(this.state.tempImage);
    this.setState({ pointer: true });
  }

  /**
 *
 * @returns {SyntheticEvent} event
 * @param {any} event
 * @memberof EditBook
 */
  onEditRequest(event) {
    event.preventDefault();
    this.props.adminModifyRequest(this.state.currentBookId, this.state);
    setTimeout(() => {
      if (this.props.modifyBookData.isModified === true) {
        this.props.history.push('/books');
        Materialize.toast('Successfully Updated', 2000, 'teal rounded');
      } else {
        Materialize.toast('Not Updated', 2000, 'red rounded');
      }
    }, 1000);
  }

  /**
     *
     *
     * @returns {ReactElement} Markup
     * @memberof EditBook
     */
  render() {
    const preview = this.state.imagePreview === '' ?
      <button
        href="#modal1"
        className="modal-trigger previewBtn"
        disabled>Preview
      </button> :
      <button
        href="#modal1"
        className="modal-trigger previewBtn">Preview
      </button>;
    const category = this.state.categoryLoader ? <div> Loading... </div> :
      <div className="row">
        <div className="input-field col s12 status">
          <select
            className="teal-text"
            id= "category"
            value="1"
            onChange={this.handleChange}>
            <option
              value={this.state.category}>
              {this.state.category}
            </option>
            { Object.keys(this.state.categoryData).map(key =>
              (<option key = {key} value={this.state.categoryData[key].category}
              >
                {this.state.categoryData[key].category }
              </option>))}
          </select>
          <label className="black-text sort">Category</label>
        </div>
      </div>;
    return (
      <div className="adminAddBooks row">
        <div className=""> <NavigationBar /> </div>
        <div className="col s12 m4 offset-m5">
          <div className="card">
            <form
              onSubmit={this.state.tempImage === '' ? this.onEditRequest :
                this.onEditCloudinaryRequest}
              id="form">
              <div className="row">
                <div className="input-field col s12">
                  <input
                    value={this.state.title}
                    onChange={this.handleChange}
                    id="title"
                    type="text"
                    className="validate"/>
                  <label className="active" htmlFor="title">Title</label>
                </div>
              </div>
              <div className="row">
                <div className="input-field col s12">
                  <input
                    value={this.state.author}
                    onChange={this.handleChange}
                    id="author"
                    type="text"
                    className="validate"/>
                  <label className="active" htmlFor="author">Author</label>
                </div>
              </div>
              { category }
              <div className="row">
                <div className="file-field input-field col s12">
                  <div className="btn file">
                    <span><i className="material-icons">file_upload</i></span>
                    <input
                      type="file"
                      onChange={this.handleImageChange}
                      id="image"/>
                  </div>
                  <div className="file-path-wrapper">
                    <input
                      value={this.state.imageView}
                      className="file-path validate"
                      type="text"
                      placeholder="Upload Image"/>
                  </div>
                </div>
                {preview}
              </div>
              <div className="row">
                <div className="input-field col s12">
                  <textarea
                    value={this.state.review}
                    onChange={this.handleChange}
                    id="review"
                    className="materialize-textarea"></textarea>
                  <label className="active" htmlFor="review">Review</label>
                </div>
              </div>
              <button
                className="btn waves-effect waves-light"
                id="submitEdit"
                type="submit"
                name="action">Update<i className="material-icons right">
                send</i>
              </button>
            </form>
            <div id="modal1" className="modal">
              <div className="modal-content">
                <img src= {this.state.imagePreview}/>
              </div>
              <div className="modal-footer">
                <a className="modal-action modal-close">
                  <button
                    className="close">Close
                  </button></a>
              </div>
            </div>
          </div>
        </div>
        <div> <Footer /></div>
      </div>
    );
  }
}

EditBook.contextTypes = {
  router: PropTypes.object.isRequired
};

export const mapStateToProps = state => (
  {
    modifyBookData: state.modifyBooks[0],
    imageInputUrl: state.uploadImage[0].response,
    getBookId: state.editBookId[0].response,
    getABookData: state.getABook[0].response,
    getCategoryData: state.getCategory[0].response,
  }
);

export default connect(
  mapStateToProps,
  {
    adminModifyRequest, getBookRequest, saveImageCloudinary, getAllCategoryRequest
  }
)(EditBook);
