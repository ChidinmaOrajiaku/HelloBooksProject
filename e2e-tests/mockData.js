module.exports = {
  file: {
    lastModified: 1508779018000,
    lastModifiedDate: 'Mon Oct 23 2017 18:16:58 GMT+0100 (WAT)',
    name: '3 card game.jpg',
    size: 642971,
    type: 'image/jpeg',
    webkitRelativePath: ''
  },
};

