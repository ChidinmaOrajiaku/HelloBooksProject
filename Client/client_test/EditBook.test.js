import React from 'react';
import $ from 'jquery';
import toJson from 'enzyme-to-json';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import mockData from './mocks/mockData';
import { EditBook } from '../components/admin/EditBook.jsx';

configure({ adapter: new Adapter() });

describe('<EditBook />', () => {
  const props = {
    getBookRequest: jest.fn(() => Promise.resolve()),
    getAllCategoryRequest: jest.fn(() => Promise.resolve()),
    saveImageCloudinary: jest.fn(),
    adminModifyRequest: jest.fn(),
    adminCountCategoryRequest: jest.fn(),
    adminCountUserRequest: jest.fn()
  };
  let mountedComponent;
  const EditBookItem = () => {
    if (!mountedComponent) {
      mountedComponent = shallow(<EditBook {...props}/>);
    }
    return mountedComponent;
  };

  it('should call onEditCloudinaryRequest method', () => {
    const spy = jest.spyOn(EditBookItem().instance(), 'onEditCloudinaryRequest');
    const event = {
      preventDefault: jest.fn(),
      target: {
        value: 'abd',
      }
    };
    EditBookItem().instance().onEditCloudinaryRequest(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should call onEditRequest method', () => {
    const spy = jest.spyOn(EditBookItem().instance(), 'onEditRequest');
    const event = {
      preventDefault: jest.fn(),
      target: {
        value: 'abd',
      }
    };
    EditBookItem().instance().onEditRequest(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should call handleChange method', () => {
    const spy = jest.spyOn(EditBookItem().instance(), 'handleChange');
    const event = {
      preventDefault: jest.fn(),
      target: {
        id: 'review',
        value: 'chideberecom'
      }
    };
    EditBookItem().instance().handleChange(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should match snapshot test', () => {
    const component = EditBookItem();
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
