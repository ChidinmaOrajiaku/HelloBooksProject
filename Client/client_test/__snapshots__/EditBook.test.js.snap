// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditBook /> should match snapshot test 1`] = `
<div
  className="adminAddBooks row"
>
  <div
    className=""
  >
     
    <Connect(withRouter(NavigationBar)) />
     
  </div>
  <div
    className="col m4 offset-m5"
  >
    <div
      className="card"
    >
      <form
        id="form"
        onSubmit={[Function]}
      >
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <input
              className="validate"
              id="title"
              onChange={[Function]}
              type="text"
              value=""
            />
            <label
              className="active"
              htmlFor="title"
            >
              Title
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <input
              className="validate"
              id="author"
              onChange={[Function]}
              type="text"
              value=""
            />
            <label
              className="active"
              htmlFor="author"
            >
              Author
            </label>
          </div>
        </div>
        <div>
           Loading... 
        </div>
        <div
          className="row"
        >
          <div
            className="file-field input-field col s12"
          >
            <div
              className="btn file"
            >
              <span>
                <i
                  className="material-icons"
                >
                  file_upload
                </i>
              </span>
              <input
                id="image"
                onChange={[Function]}
                type="file"
              />
            </div>
            <div
              className="file-path-wrapper"
            >
              <input
                className="file-path validate"
                placeholder="Upload Image"
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            className="modal-trigger previewBtn"
            disabled={true}
            href="#modal1"
          >
            Preview
          </button>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <textarea
              className="materialize-textarea"
              id="review"
              onChange={[Function]}
              value=""
            />
            <label
              className="active"
              htmlFor="review"
            >
              Review
            </label>
          </div>
        </div>
        <button
          className="btn waves-effect waves-light"
          id="submitEdit"
          name="action"
          type="submit"
        >
          Update
          <i
            className="material-icons right"
          >
            send
          </i>
        </button>
      </form>
      <div
        className="modal"
        id="modal1"
      >
        <div
          className="modal-content"
        >
          <img
            src=""
          />
        </div>
        <div
          className="modal-footer"
        >
          <a
            className="modal-action modal-close"
          >
            <button
              className="close"
            >
              Close
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div>
     
    <Component />
  </div>
</div>
`;
